File: app/src/main/java/tech/techlore/plexus/activities/SplashActivity.java
/*
 * Copyright (c) 2022 Techlore
 *
 *  This file is part of Plexus.
 *
 *  Plexus is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Plexus is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Plexus.  If not, see <https://www.gnu.org/licenses/>.
 */

package tech.techlore.plexus.activities;

import static tech.techlore.plexus.utils.IntentUtils.SendListsIntent;
import static tech.techlore.plexus.utils.NetworkUtils.HasInternet;
import static tech.techlore.plexus.utils.NetworkUtils.HasNetwork;
import static tech.techlore.plexus.utils.NetworkUtils.GETReq;
import static tech.techlore.plexus.utils.ListUtils.PopulateDataList;
import static tech.techlore.plexus.utils.ListUtils.ScanInstalledApps;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

import tech.techlore.plexus.R;
import tech.techlore.plexus.models.InstalledApp;
import tech.techlore.plexus.models.PlexusData;

@SuppressLint("CustomSplashScreen")
public class SplashActivity extends AppCompatActivity {

    private List<PlexusData> plexusDataList;
    private List<InstalledApp> installedAppsList;
    private static String jsonData;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash);

        plexusDataList = new ArrayList<>();
        installedAppsList = new ArrayList<>();

        /*###########################################################################################*/

        FetchData();

    }

    private void NoNetworkDialog() {

        new MaterialAlertDialogBuilder(this, R.style.DialogTheme)

                .setTitle(R.string.dialog_title)
                .setMessage(R.string.dialog_subtitle)

                .setPositiveButton(R.string.retry, (dialog, which) ->
                        FetchData())

                .setNegativeButton(R.string.exit, (dialog, which) ->
                        finishAndRemoveTask())
                
                .setCancelable(false)

                .show();

    }

    private void FetchData(){

        Handler handler = new Handler(Looper.getMainLooper());

        if (HasNetwork(this)) {

            Executors.newSingleThreadExecutor().execute(() -> {

                // Background thread work
                if (HasInternet()) {

                    try {
                        jsonData = GETReq();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }

                    // UI Thread work
                    handler.post(() -> {
                        try {
                            plexusDataList = PopulateDataList(jsonData);
                            ((TextView)findViewById(R.id.progress_text)).setText(R.string.scan_installed);
                            ScanInstalledApps(this, plexusDataList, installedAppsList);
                            SendListsIntent(this, MainActivity.class,
                                    (Serializable) plexusDataList, (Serializable) installedAppsList);
                            finish();
                            overridePendingTransition(R.anim.slide_from_end, R.anim.slide_to_start);
                        }

                        catch (JsonProcessingException e) {
                            e.printStackTrace();
                        }
                    });
                }
                else {
                    handler.post(this::NoNetworkDialog);
                }
            });
        }
        else {
            NoNetworkDialog();
        }

    }
}

File: app/src/main/java/tech/techlore/plexus/activities/SplashActivity.kt
/*
 * Copyright (c) 2022-present Techlore
 *
 *  This file is part of Plexus.
 *
 *  Plexus is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Plexus is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Plexus.  If not, see <https://www.gnu.org/licenses/>.
 */

package tech.techlore.plexus.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import tech.techlore.plexus.R
import tech.techlore.plexus.appmanager.ApplicationManager
import tech.techlore.plexus.databinding.ActivitySplashBinding
import tech.techlore.plexus.fragments.bottomsheets.FirstLaunchBottomSheet
import tech.techlore.plexus.fragments.bottomsheets.NoNetworkBottomSheet
import tech.techlore.plexus.preferences.PreferenceManager
import tech.techlore.plexus.preferences.PreferenceManager.Companion.IS_FIRST_LAUNCH
import tech.techlore.plexus.utils.NetworkUtils.Companion.hasInternet
import tech.techlore.plexus.utils.NetworkUtils.Companion.hasNetwork
import tech.techlore.plexus.utils.SystemUtils.Companion.isDeviceDeGoogledOrMicroG

@SuppressLint("CustomSplashScreen")
class SplashActivity : AppCompatActivity() {
    
    lateinit var activityBinding: ActivitySplashBinding
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityBinding = ActivitySplashBinding.inflate(layoutInflater)
        setContentView(activityBinding.root)
        
        retrieveData()
    }
    
    private fun retrieveData() {
        lifecycleScope.launch{
            val context = this@SplashActivity
            if (hasNetwork(context) && hasInternet()) {
                val mainRepository = (applicationContext as ApplicationManager).mainRepository
                mainRepository.plexusDataIntoDB(context)
                activityBinding.progressText.text = getString(R.string.scan_installed)
                mainRepository.installedAppsIntoDB(context)
                isDeviceDeGoogledOrMicroG(context)
                afterDataRetrieved()
            }
            else {
                NoNetworkBottomSheet(negativeButtonText = getString(R.string.exit),
                                     positiveButtonClickListener = { retrieveData() },
                                     negativeButtonClickListener = { finishAndRemoveTask() })
                    .show(supportFragmentManager, "NoNetworkBottomSheet")
            }
        }
    }
    
    private fun afterDataRetrieved() {
        if (PreferenceManager(this@SplashActivity).getBoolean(IS_FIRST_LAUNCH)) {
            FirstLaunchBottomSheet().show(supportFragmentManager, "FirstLaunchBottomSheet")
        }
        else {
            startActivity(Intent(this@SplashActivity, MainActivity::class.java))
            finish()
        }
        activityBinding.progressIndicator.apply {
            pauseAnimation()
            isVisible = false
        }
        activityBinding.progressText.isVisible = false
    }
    
    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.slide_from_end, R.anim.slide_to_start)
    }
}